File: app/adapters/interaction_rule/mappers/interaction_rule_agrr_mapper.rb
# frozen_string_literal: true

module Adapters
  module InteractionRule
    module Mappers
      class InteractionRuleAgrrMapper
        def self.to_agrr_format(entity_or_record)
          record = entity_or_record
          id = record.respond_to?(:id) ? record.id : record[:id]
          {
            "rule_id" => "rule_#{id}",
            "rule_type" => fetch_attr(record, :rule_type),
            "source_group" => fetch_attr(record, :source_group),
            "target_group" => fetch_attr(record, :target_group),
            "impact_ratio" => fetch_attr(record, :impact_ratio).to_f,
            "is_directional" => fetch_attr(record, :is_directional),
            "description" => fetch_attr(record, :description)
          }.compact
        end

        def self.to_agrr_format_array(entities_or_records)
          Array(entities_or_records).map { |r| to_agrr_format(r) }
        end

        def self.fetch_attr(record, key)
          if record.respond_to?(key)
            record.public_send(key)
          else
            record[key] || record[key.to_s]
          end
        end

        private_class_method :fetch_attr
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/mappers/interaction_rule_agrr_mapper.rs
//! Ruby: `Adapters::InteractionRule::Mappers::InteractionRuleAgrrMapper`

use serde_json::{json, Value};

use crate::interaction_rule::entities::InteractionRuleEntity;

pub fn to_agrr_format(rule: &InteractionRuleEntity) -> Value {
    let id = rule.id.unwrap_or(0);
    let mut obj = serde_json::Map::new();
    obj.insert("rule_id".into(), json!(format!("rule_{id}")));
    obj.insert("rule_type".into(), json!(rule.rule_type));
    obj.insert("source_group".into(), json!(rule.source_group));
    obj.insert("target_group".into(), json!(rule.target_group));
    obj.insert("impact_ratio".into(), json!(rule.impact_ratio));
    if let Some(v) = rule.is_directional {
        obj.insert("is_directional".into(), json!(v));
    }
    if let Some(ref d) = rule.description {
        obj.insert("description".into(), json!(d));
    }
    Value::Object(obj)
}

pub fn to_agrr_format_array(rules: &[InteractionRuleEntity]) -> Vec<Value> {
    rules.iter().map(to_agrr_format).collect()
}
